sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java

```diff
@@ -162,6 +201,106 @@ public Duration getOperationTimeout() {
+        }
+
+        return RetryUtil.withRetry(
+            getEndpointStates().takeUntil(state -> state == AmqpEndpointState.ACTIVE),
+            timeout, retry)
+            .then(Mono.<AmqpLink>create(sink -> {
+                try {
+                    // We have to invoke this in the same thread or else proton-j will not properly link up the created
+                    // sender because the link names are not unique. Link name == entity path.
+                    provider.getReactorDispatcher().invoke(() -> {
+                        LinkSubscription<AmqpLink> linkLinkSubscription = getCoordinator(linkName, timeout, retry);
+
+                        if (coordinator.compareAndSet(null, linkLinkSubscription)) {
+                            logger.info("linkName[{}]: coordinator link created.", linkName);
+                        } else {
+                            logger.info("linkName[{}]: Another coordinator link exists. Disposing of new one.",
+                                linkName);
+                            linkLinkSubscription.dispose();
+                        }
+
+                        sink.success(coordinator.get().getLink());
+                    });
+                } catch (IOException e) {
+                    sink.error(e);
+                }
+            }));
+    }
+
+    /**
+     * NOTE: Ensure this is invoked using the reactor dispatcher because proton-j is not thread-safe.
+     */
+    private LinkSubscription<AmqpLink> getCoordinator(String linkName, Duration timeout, AmqpRetryPolicy retry) {
+
+        final Sender sender = session.sender(linkName);
+        sender.setTarget(new Coordinator());
+
+        final Source source = new Source();
+        sender.setSource(source);
+        sender.setSenderSettleMode(SenderSettleMode.UNSETTLED);
+
+        final SendLinkHandler sendLinkHandler = handlerProvider.createSendLinkHandler(
+            sessionHandler.getConnectionId(), sessionHandler.getHostname(), linkName, linkName);
+        BaseHandler.setHandler(sender, sendLinkHandler);
+
+        sender.open();
+
+        final ReactorSender coordinator = new ReactorSender(linkName, sender, sendLinkHandler, provider, null,
+            messageSerializer, timeout, retry);
+
+        final Disposable subscription = coordinator.getEndpointStates().subscribe(state -> { },
+            error -> {
+                logger.info("linkName[{}]: Error occurred. Removing and disposing coordinator link.", linkName, error);
+                removeLink(openSendLinks, linkName);
+            }, () -> {
+                logger.info("linkName[{}]: Complete. Removing and disposing coordinator link.", linkName);
+                removeLink(openSendLinks, linkName);
+            });
+        return new LinkSubscription<>(coordinator, subscription);
+    }
+
+    public Mono<AmqpTransaction> createTransaction() {
+        return createTransactionCoordinator()
+            .cast(ReactorSender.class)
```

Comment:
I'm wary of casting this to ReactorSender. Does the TransactionCoordinator also have a ReactorSender? Your current implementation adds more responsibility to ReactorSender. It not only handles sending messages, but also transactions. It would be a better separation of concerns to have a TransactionCoordinator class that contains a AmqpSendLink that verifies send operations of the "transaction" message.


```java
class TransactionCoordinator {
    private final AmqpSendLink sendLink;
    TransactionCoordinator(AmqpSendLink link) {
        // This could be reactor sender underneath.
        this.sendLink = link;
    }
}

Reply:
I already had a separate class `ReactorCoordinator` , in order to separate concern butI had to replicate most of the code from `ReactorSender`  into this. You pointed that out in  your comment https://github.com/Azure/azure-sdk-for-java/pull/11305#discussion_r431517190   . Actually Transaction/Coordinator a logical name, we are actually sending Amqp Message over `AmqpSendLink ` . We are sending Amqp Message with different properties to create transactions.  So I think keeping all this in ReactorSender make sense, otherwise we have to replicate 95% same logic from `ReactorSender` in whatever new class we make.